MicroProfile OpenAPI 4.1 Implementation

Azquelt opened the pull request:
Add the MicroProfile OpenAPI 4.1 Implementation

- Update the API to 4.1.0-RC1
- Update the SmallRye OpenAPI implementation to 4.1.0-alpha3
- Add a new project to run the 4.1-RC1 TCK
- Refactor the 4.0 implementation into the parts that are common and the parts that need to change for 4.1
- Implement `ModelOperations.shallowCopy` for 4.1, adding in the copy of the new `jsonSchemaDialect` property

Fixes #31298 
Fixes #31303 
Fixes #31300 

- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Azquelt committed:
Update MP OpenAPI 4.1 API to RC1

Azquelt committed:
Run MP OpenAPI 4.1-RC1 TCK

Azquelt committed:
Use SmallRye OpenAPI 4.1.0-alpha3

Azquelt committed:
Update bundles in mpOpenAPI-4.1 feature

Azquelt committed:
Split common parts of MP OpenAPI 4.0 impl

Split up the MP OpenAPI 4.0 implementation classes into the ones that
will be shared with the 4.1 impl and the ones that won't.

Azquelt committed:
Create mpOpenAPI-4.1 specific implementation

Remove MP OpenAPI 4.1 test project since the tests can now be included
in the new implementation bundle project.

Azquelt commented on dev/io.openliberty.microprofile.openapi.4.0.internal/bnd.bnd:
```diff
@@ -0,0 +1,68 @@
+#*******************************************************************************
+# Copyright (c) 2024, 2025 IBM Corporation and others.
```
Copied from `io.openliberty.microprofile.openapi.4.0.internal.services/bnd.bnd` since most of the code from there has been moved to this new project.

Azquelt commented on dev/io.openliberty.microprofile.openapi.4.0.internal/src/io/openliberty/microprofile/openapi40/internal/impl/OpenAPI31ModelOperations.java:
```diff
@@ -33,7 +30,6 @@
 import io.smallrye.openapi.runtime.io.JsonIO;
 
-@Component(configurationPolicy = ConfigurationPolicy.IGNORE, service = OpenAPIModelOperations.class)
 public class OpenAPI31ModelOperations extends OpenAPIModelOperationsImpl {
```
This is the main reason for refactoring this code into a common project for 4.0 and 4.1. We need a slightly different implementation of this service in 4.1.

There are two new classes in the other bundles which extend this class and declare a component.

Azquelt commented on dev/io.openliberty.microprofile.openapi.4.0.internal.services/src/io/openliberty/microprofile/openapi40/internal/services/impl/package-info.java:
```diff
@@ -1,13 +1,15 @@
 /*******************************************************************************
- * Copyright (c) 2024, 2025 IBM Corporation and others.
+ * Copyright (c) 2025 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
  *
  * SPDX-License-Identifier: EPL-2.0
  *******************************************************************************/
-@com.ibm.websphere.ras.annotation.TraceOptions(traceGroup = Constants.TRACE_GROUP, messageBundle = Constants.TRACE_OPENAPI)
+@TraceOptions(traceGroup = Constants.TRACE_GROUP)
 package io.openliberty.microprofile.openapi40.internal.services.impl;
```
This slightly odd change is because the original package was refactored into a different project and a new package created with the same name.

Azquelt commented on dev/io.openliberty.microprofile.openapi.4.1.internal.services/src/io/openliberty/microprofile/openapi41/internal/services/impl/OpenAPI31ModelOperations41Impl.java:
```diff
@@ -0,0 +1,28 @@
+@Component(configurationPolicy = ConfigurationPolicy.IGNORE, service = OpenAPIModelOperations.class)
+public class OpenAPI31ModelOperations41Impl extends OpenAPI31ModelOperations {
+
+    @Override
+    public OpenAPI shallowCopy(OpenAPI model) {
+        OpenAPI copy = super.shallowCopy(model);
+        copy.setJsonSchemaDialect(model.getJsonSchemaDialect());
+        return copy;
+    }
```
All of the refactoring is just to add this method :(

Azquelt commented on dev/io.openliberty.microprofile.openapi.4.1.internal_fat_tck/build.gradle:
```diff
@@ -0,0 +1,17 @@
+  requiredLibs 'org.hamcrest:hamcrest:2.1'
+}
+
+addRequiredLibraries.dependsOn addJakartaTransformer
```
Can remove this when #31324 is done

jhanders34 replied:
Do we use fattest.simplicity to install applications?  If no, is it really needed?

Azquelt replied:
Good point, it shouldn't be.

Azquelt commented on dev/io.openliberty.microprofile.openapi.4.1.internal_fat_tck/fat/src/io/openliberty/microprofile/openapi41/fat/tck/OpenAPITckTest.java:
```diff
@@ -0,0 +1,73 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2025 IBM Corporation and others.
```
Almost a direct copy of the TCK test class for the previous version.

Azquelt commented on dev/io.openliberty.microprofile.openapi.4.1.internal_fat_tck/publish/tckRunner/tck/src/test/java/io/openliberty/microprofile/openapi40/test/ArquillianLoadableExtension.java:
```diff
@@ -0,0 +1,28 @@
+/*******************************************************************************
+ * Copyright (c) 2017, 2024 IBM Corporation and others.
```
Copied unchanged

Azquelt commented:
!build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=3986df51-de5c-4919-8e02-6e0de3475a44) - :x: completed with errors/failures.)
!spawn.fullfat.buckets=com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.openapi.ui_fat,io.openliberty.microprofile.openapi.2.0.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat,io.openliberty.microprofile.openapi.ui.internal_fat

_Note: Target locations of links might be accessible only to IBM employees._

jhanders34 commented on dev/io.openliberty.microprofile.openapi.4.0.internal.services/src/io/openliberty/microprofile/openapi40/internal/services/impl/package-info.java:
```diff
@@ -1,13 +1,15 @@
 /*******************************************************************************
- * Copyright (c) 2024, 2025 IBM Corporation and others.
+ * Copyright (c) 2025 IBM Corporation and others.
```
```suggestion
 * Copyright (c) 2024, 2025 IBM Corporation and others.
```

Azquelt replied:
The original file was refactored into the other bundle and this package created fresh (albeit with the same name) which is why the copyright date was not retained.

I guess the name is still the same so you could argue it should stay? Unsure.

Azquelt replied:
Changed it back

jhanders34 commented on dev/io.openliberty.microprofile.openapi.4.0.internal/src/io/openliberty/microprofile/openapi40/internal/impl/package-info.java:
```diff
@@ -0,0 +1,14 @@
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+@com.ibm.websphere.ras.annotation.TraceOptions(traceGroup = Constants.TRACE_GROUP, messageBundle = Constants.TRACE_OPENAPI)
+@org.osgi.annotation.versioning.Version("0.1")
```
```suggestion
@org.osgi.annotation.versioning.Version("1.0")
```
First version of an exported package is usually 1.0

jhanders34 commented on dev/io.openliberty.microprofile.openapi.4.0.internal/bnd.bnd:
```diff
@@ -0,0 +1,68 @@
+    *
+
+Export-Package: \
+    io.openliberty.microprofile.openapi40.internal.impl
```
Usually packages with `.internal.` in the name are not exported.   Do we need to remove `.internal` from the package name?

Azquelt replied:
It's internal implementation detail which is exported because this bundle is shared between 4.0 and 4.1. Other components shouldn't be using it and it's not API or SPI.

From the [naming guidelines](https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-naming-guidelines) I think I should be using `internal` in the name for this case.

jhanders34 commented on dev/io.openliberty.microprofile.openapi.4.1.internal_fat_tck/bnd.bnd:
```diff
@@ -0,0 +1,34 @@
+
+tested.features: \
+  validation-3.1,\
+  expressionlanguage-5.0,\
+  expressionlanguage-6.0,\
+  restfulWS-3.0,\
+  restfulWS-4.0,\
+  servlet-6.0,\
+  servlet-7.0
```
```suggestion
  restfulWS-4.0
```
I don't believe you need the other ones.  restfulWS-3.0 is EE 9 and definitely wouldn't be used and servlet-7.0 isn't a thing.  For EE 11 it is servlet-6.1, but I don't think you use servlet for these tests.

Azquelt replied:
I think you're right, this list looks like it hasn't been updated.

Azquelt replied:
Test complains if el-5.0 is removed so I put it back. Not sure why, I assume there's some tolerates going on.

jhanders34 replied:
It makes sense.  beanValidation-3.0 is EE 9 and 10 so when it looks at it looks at the preferred for EL which is 4.0 so 5.0 needs to be added.

jhanders34 commented on dev/io.openliberty.org.eclipse.microprofile/openapi.4.1.bnd:
```diff
@@ -19,9 +19,9 @@ Export-Package: \
 
 -maven-dependencies: \
-   dep1;groupId=org.eclipse.microprofile.openapi;artifactId=microprofile-openapi-api;version=4.0.2;scope=runtime
+   dep1;groupId=org.eclipse.microprofile.openapi;artifactId=microprofile-openapi-api;version=4.1;scope=runtime
```
Does this need to be precise i.e. 4.1-RC1 for the scanning function to find the resource?

Azquelt replied:
Possibly? I'm not sure what this is actually used for...

Azquelt replied:
Changed it in case it is important.

jhanders34 replied:
If we used an entire project instead of using a sub bundle bnd file, the buildpath would have openapi-api:4.1-RC1 so it would add that to the pom of the jar for the scanning tooling to see all dependencies that were used for the bundle.  With using sub bundles without a buildpath, `-maven-dependencies` is how we add dependencies to the bundle's pom file.

Azquelt committed:
Trim list of affected features for MP OpenAPI 4.1 TCK FAT

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

Azquelt committed:
Fix bundle export version

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

Azquelt committed:
Remove TCK FAT transformer dependency

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

Azquelt committed:
List the MP OpenAPI 4.1 API artifact accurately

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

Azquelt committed:
Fix copyright date

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

Azquelt commented:
!build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=40334fbc-e9a2-4630-a75e-9e467b4279ad) - :x: completed with errors/failures.)
!spawn.fullfat.buckets=com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.openapi.ui_fat,io.openliberty.microprofile.openapi.2.0.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat,io.openliberty.microprofile.openapi.ui.internal_fat

_Note: Target locations of links might be accessible only to IBM employees._

Azquelt commented:
Need to double-check the smallrye jars have been rebundled correctly before merging.

Azquelt committed:
Use a range with repo macro for SR OpenAPI modules

When including resources using the ${repo} macro, bnd will treat a
version as a lower bound with no upper bound and choose the highest
matching version to satisfy the dependency.

Since we don't want any new minor version to be picked up by these
projects, we need to specify either a range with an upper bound or use
the EXACT strategy.

Azquelt commented:
Inspecting the classes for the 4.0 bundles they do match the upstream jar, but in the case the jaxrs module I think this is only because there have been no upstream changes to that module.

I've updated these references with a bounded range rather than a plain version (which is treated as a lower bound).

Azquelt commented:
!build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=53e04b71-d779-4d35-af87-777d16ead142) - :white_check_mark: completed successfully!)
spawn.fullfat.buckets=com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.openapi.ui_fat,io.openliberty.microprofile.openapi.2.0.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat,io.openliberty.microprofile.openapi.ui.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat_tck,io.openliberty.microprofile.openapi.4.1.internal_fat_tck

_Note: Target locations of links might be accessible only to IBM employees._

Azquelt commented:
Manually checked the classes in the 4.0 bundles match the upstream sources again.
